### PR TITLE
[IDP-1772] Display Spectral report in ADO CI build extension

### PR DESCRIPTION
**Pull request by heqianwang**
## Description of changes
We want to display our changes as an ADO pipeline extension automatically instead of manually it as a pipeline step. Basically, when we are done generating a spectral report, we display it using a console command.

## Breaking changes
Users relying on HTML reports will find that their reports are now in Txt format given that ADO is not compatible with HTML..
## Additional checks
Check with a preview version of the library.
image

### CODE REVIEW

File: src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs [hunk 1]
```diff
@@ -161,4 +177,16 @@ private async Task GeneratePublicNugetSource()
             File.WriteAllText(path, "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n<configuration>\n  <packageSources>\n    <clear />\n    <add key=\"nuget\" value=\"https://api.nuget.org/v3/index.json\" />\n  </packageSources>\n</configuration>");
         }
     }
+
+    // In the future, we will want to support both ADO and Github CI environments
+    private AdoCiReportRenderer InitializeCiReportRenderer()
+    {
```
Comment (heqianwang) [inline src/Workleap.OpenApi.MSBuild/ValidateOpenApiTask.cs:183]: Right, we only support ADO environments so I've left this as AdoCiReportRenderer return value by default but once we migrate and support more products in Github, will update this accordingly

File: README.md [hunk 2]
```diff
@@ -17,7 +17,8 @@ For the TLDR version:
...
 - The default value are defined in the property group on the target `ValidateOpenApi` in this file `./src/Workleap.OpenApi.MSBuild/msbuild/tools/Workleap.OpenApi.MSBuild.targets`
-- Users can select whether to validate the API with frontend or backend ruleset depending how they configure the `OpenApiServiceProfile` MSBuild property ([`backend` (default)](https://github.com/gsoft-inc/wl-api-guidelines/blob/main/.spectral.backend.yaml) or [`frontend`](https://github.com/gsoft-inc/wl-api-guidelines/blob/main/.spectral.frontend.yaml)).
+- Users can select whether to validate the API with frontend or backend ruleset depending on how they configure the `OpenApiServiceProfile` MSBuild property ([`backend` (default)](https://github.com/gsoft-inc/wl-api-guidelines/blob/main/.spectral.backend.yaml) or [`frontend`](https://github.com/gsoft-inc/wl-api-guidelines/blob/main/.spectral.frontend.yaml)).
+- Users can configure their CI enviroment for reports depending on how they configure the `OpenApiCiReportEnvironment` MSBuild property (`ado` (default)).
```
Comment (asimmon) [inline README.md:21]: Do we need the developer to provide this information? We can detect whether or not the build is running on ADO using well-known environment variables, like we already do here:

https://github.com/gsoft-inc/wl-extensions-xunit/blob/bd7d883917a1915b5af54d4fe848594bc4dab1ca/src/Workleap.Extensions.Xunit/BaseIntegrationFixture.cs#L77-L79

File: src/Workleap.OpenApi.MSBuild/Spectral/AdoCiReportRenderer.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+namespace Workleap.OpenApi.MSBuild.Spectral;
+
+internal sealed class AdoCiReportRenderer : ICiReportRenderer
+{
+    public async Task AttachReportToBuildAsync(string reportPath)
+    {
+        // Attach the report to the build summary
+        if (!string.IsNullOrEmpty(Environment.GetEnvironmentVariable("AGENT_NAME")))
```
Comment (asimmon) [inline src/Workleap.OpenApi.MSBuild/Spectral/AdoCiReportRenderer.cs:8]: `AGENT_NAME` might not be specific enough to Azure DevOps. [We detect](https://github.com/gsoft-inc/wl-extensions-xunit/blob/bd7d883917a1915b5af54d4fe848594bc4dab1ca/src/Workleap.Extensions.Xunit/BaseIntegrationFixture.cs#L77-L79) `SYSTEM_TEAMFOUNDATIONCOLLECTIONURI` (Team Foundation = previous name of Azure DevOps, which used to be called VSTS at some point too) in this project and it was based on a library that is specialized in CI detection.

File: README.md (hunk 2 above)
Comment (asimmon) [reply README.md:21]: Also I don't understand why we sometimes use this user-provided information, then sometimes look up for an environment variable

File: src/Workleap.OpenApi.MSBuild/Spectral/AdoCiReportRenderer.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
+namespace Workleap.OpenApi.MSBuild.Spectral;
+
+internal sealed class AdoCiReportRenderer : ICiReportRenderer
+{
+    public async Task AttachReportToBuildAsync(string reportPath)
```
Comment (asimmon) [inline src/Workleap.OpenApi.MSBuild/Spectral/AdoCiReportRenderer.cs:5]: This method doesn't seem to be asynchronous

**Comment by heqianwang** [comment]
Reran the E2E scenario and both console & extensions outputs are valid: 
image
image

File: README.md (hunk 2 above)
Comment (heqianwang) [reply README.md]: Discussed offline and full agree with that, we'll default to using environment variables present on the CI build env

File: src/Workleap.OpenApi.MSBuild/Spectral/CiReportRenderer.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
+namespace Workleap.OpenApi.MSBuild.Spectral;
+
+internal static class CiReportRenderer
+{
+    public static void AttachReportToBuildAsync(string reportPath)
```
[the remark this answers is not in the document]
Comment (asimmon) [reply src/Workleap.OpenApi.MSBuild/Spectral/CiReportRenderer.cs]:
```suggestion
    public static void AttachReportToBuild(string reportPath)
```

File: src/Workleap.OpenApi.MSBuild/Spectral/SpectralRunner.cs [hunk 6]
```diff
@@ -50,19 +50,20 @@ public async Task RunSpectralAsync(IReadOnlyCollection<string> openApiDocumentPa
...
             {
-                this._loggerWrapper.LogMessage("\nDeleting existing report: {0}", messageArgs: htmlReportPath);
-                File.Delete(htmlReportPath);
+                this._loggerWrapper.LogMessage("\nDeleting existing report: {0}", messageArgs: spectralReportPath);
+                File.Delete(spectralReportPath);
             }
             
-            await this.GenerateSpectralReport(spectralExecutePath, documentPath, spectralRulesetPath, htmlReportPath, cancellationToken);
+            await this.GenerateSpectralReport(spectralExecutePath, documentPath, spectralRulesetPath, spectralReportPath, cancellationToken);
+            CiReportRenderer.AttachReportToBuildAsync(spectralReportPath);
```
Comment (asimmon) [inline src/Workleap.OpenApi.MSBuild/Spectral/SpectralRunner.cs]:
```suggestion
            CiReportRenderer.AttachReportToBuild(spectralReportPath);
```

**Review by asimmon — CHANGES_REQUESTED** [review]

File: src/Workleap.OpenApi.MSBuild/Spectral/SpectralRunner.cs (hunk 6 above)
Comment (heqianwang) [reply src/Workleap.OpenApi.MSBuild/Spectral/SpectralRunner.cs]: woops

**Review by asimmon — APPROVED** [review]

**Pull request merged**